### ISSUE TITLE
The source_xpos parameter not being calculated for FS data?

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#6074]
_Issue [JP-2103](https://jira.stsci.edu/browse/JP-2103) was created on JIRA by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol):_

In processing a set of FS simulated dithered exposures, which included both primary and spectral dithers, both the source_xpos and source_ypos parameters are zero for all dither positions.  The former in particular should be nonzero for dither positions in which the source is not at the center of the slit in the dispersion direction, which affects the wavecorr step.  In this case, no wavelength corrections are applied even in cases where they should be.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=529654#comment-529654):_

For NIRSpec MOS mode the source_xpos/source_ypos values are retrieved from the meta data supplied by the MSA Planning Tool. How would they be calculated (on-the-fly) for FS mode? Just use the TARG_RA/TARG_DEC info to compute a location in the image, relative to where we think the center of the slit is located? Or perhaps use the x_offset/y_offset values in some way?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=529674#comment-529674):_

Either one could be used, though the former would currently be subject to the bug we recently found (JP-2083).

I'm a bit confused, though, because the documentation says this has already been implemented:  "Fixed slit data do not have an *a priori* estimate of the source location within a given slit, so the estimated source location is computed by the `wavecorr` step. It uses the target coordinates in conjunction with the aperture reference point in V2/V3 space to estimate the fractional location of the source within the given slit."

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=540285#comment-540285):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) After looking at the code (and running it) I see that the wavecorr code internally computes source_(x/y)pos and the corresponding wavelength correction but does not update the SRCXPOS keyword.

The calculation is based on the meta.dither.x(y)_offset (FITS: XOFFSET/YOFFSET) keywords and is done here

<https://github.com/spacetelescope/jwst/blob/master/jwst/wavecorr/wavecorr.py#L225>


If we are to update the XOFFSET keyword in what frame should we report the computed values? The corresponding MOS positions are in a frame where (0, 0) is at the upper left corner and (1,1) is the lower right corner. Is this how we should report them for FS?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=540378#comment-540378):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) the X/YOFFSET keywords should be in the frame where 0,0 is the center of the slit and 0.5,0.5 is the top right-hand corner.  The same should be true for MOS after conversion of the values from the PPSDB (e.g., <https://jira.stsci.edu/browse/JP-1151).>

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=541205#comment-541205):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) I needed to refresh my memory. Here's how this is implemented for the FS.

Wavecorr works on the primary slit only. It computes the location of the source in the aperture, i.e. on the MSA. If this is the "S400A1" slit it computes the fractional position within this slit on the MSA. To do this it takes "meta.dither.x_offset" and "meta.dither.y_offset" (which are in the Ideal frame), transforms them first to V2V3, and then to the "msa_frame". The fractional position is computed using the size of the aperture on the MSA.

The interpolation is done using computed lambda and fractional position in x.

I've updated the code to compute also the source position in the slit and report it in "slit.x_offset" and "slit.y_offset".

Does this sound correct?

#6243

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=541302#comment-541302):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) Yes, that looks right.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=541446#comment-541446):_

Fixed by #6243

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=542650#comment-542650):_

Also #6261 added to make code more robust against missing keywords

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=542651#comment-542651):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) and yet another one ready for testing

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=544063#comment-544063):_

The keywords are now being populated, however, the values are not correct. For a set of spectral dithers taken with the S200A1 slit, I see values of [-0.29, -0.42, and -0.54]; the expected values should be roughly [-0.26, 0.0, 0.26], based on the pattern positions [-0.0264, 0, 0.0264] in the ideal frame and a slit width of 0.2 arcsec. My guess is that the wrong aperture reference point is being used, and also a factor of 2 is missing (i.e., the slit edges should be +/- 0.5, not 1). SRCYPOS has the correct relative sequence, but again is missing a factor of 2:  the calculated values are ~ [-0.36, -0.03, 0.23], but should be ~ [-0.66, 0, 0.53] based on the 3 pt primary dither pattern positions of [1.0898, 0, -0.879] in the ideal frame, and a slit length of 3.3 arcsec.

It also appears that no wavelength correction is actually being applied regardless of the SRCXPOS value, but I will submit a separate ticket for that.

**Comment by nden** [comment]
Which three files in the above directory do those values correspond to?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=544095#comment-544095):_

The dither position is indicated by the _d0* suffix, in this order (horizontal represents the spectral dithers, vertical the primary dithers):

d01 d02 d03

d04 d05 d06

d07 d08 d09

So, d01/d04/d07 have roughly the same SRCXPOS, and likewise for the rest.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=558872#comment-558872):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) PR #6457 addresses some of the issues we discussed, also mentioned above. There's one remaining issue - the sign of the x position in the slit is opposite of the sign of meta.dither.x_offset. For example,
```java

x_offset = -0.0264 : 
v2, v3 for position #1: 332.8346289845707, -480.061102906561
mx, my in the MSA frame for position #1: 0.0269827684100399, -0.0031764409819799184, 2e-06
sx, sy in the slit frame for position #1: 0.12711058026410074, -0.3616656933616285, 2e-06


xoffset = 0.0264 : 
v2, v3 for position #1: 332.874332560069, -480.02629678507844
mx, my in the MSA frame for position #1: 0.026962084764319105, -0.0031766109926512263, 2e-06
sx, sy in the slit frame for position #1: -0.12714443474453552, -0.3617994368213751, 2e-06

```
When we discussed this I removed "vparity" from the transform from Ideal to V2V3 frame and that "fixed" the sign issue. But I realized that this is incorrect and the parity needs to be taken into account. 

 

So my question is where in the instrument model should the parity be included - is this the OTE reference file (transform)? Since these are two different ways of computing V2, V3 coordinates, one using SIAF transforms, one using the instrument model, it seems to me there's no agreement between them in terms of which way the rotation goes. This is just a guess on my part so it may be completely off.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559192#comment-559192):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) After mulling this over to the point of smoke rising from my head, I think I may have been misinterpreting the definition of x_/y_offset. If those represent the motion of the telescope during the dither, as opposed to the source position, then the sign should be opposite. In that case, vparity should be included in the ideal-v2v3 transform, and we should be ok. Before you proceed, let me verify with the PPS folks, since the values are originally populated in the PPSDB by APT.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559400#comment-559400):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva)  I confirm that x_/y_offset indicates the telescope motion, so the sign should be opposite that of the source position within an aperture.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559425#comment-559425):_

🎉 🎉 🎉 I hope it's all good this time. Here's the PR

#6457

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559444#comment-559444):_

Fixed by #6457

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559604#comment-559604):_

Tested on simulated FS data, x and ypos values are now being calculated and stored correctly.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=529654#comment-529654):_

For NIRSpec MOS mode the source_xpos/source_ypos values are retrieved from the meta data supplied by the MSA Planning Tool. How would they be calculated (on-the-fly) for FS mode? Just use the TARG_RA/TARG_DEC info to compute a location in the image, relative to where we think the center of the slit is located? Or perhaps use the x_offset/y_offset values in some way?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=529674#comment-529674):_

Either one could be used, though the former would currently be subject to the bug we recently found (JP-2083).

I'm a bit confused, though, because the documentation says this has already been implemented:  "Fixed slit data do not have an *a priori* estimate of the source location within a given slit, so the estimated source location is computed by the `wavecorr` step. It uses the target coordinates in conjunction with the aperture reference point in V2/V3 space to estimate the fractional location of the source within the given slit."

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=540285#comment-540285):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) After looking at the code (and running it) I see that the wavecorr code internally computes source_(x/y)pos and the corresponding wavelength correction but does not update the SRCXPOS keyword.

The calculation is based on the meta.dither.x(y)_offset (FITS: XOFFSET/YOFFSET) keywords and is done here

<https://github.com/spacetelescope/jwst/blob/master/jwst/wavecorr/wavecorr.py#L225>


If we are to update the XOFFSET keyword in what frame should we report the computed values? The corresponding MOS positions are in a frame where (0, 0) is at the upper left corner and (1,1) is the lower right corner. Is this how we should report them for FS?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=541302#comment-541302):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) Yes, that looks right.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=541446#comment-541446):_

Fixed by #6243

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=542650#comment-542650):_

Also #6261 added to make code more robust against missing keywords

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=542651#comment-542651):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) and yet another one ready for testing

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=544063#comment-544063):_

The keywords are now being populated, however, the values are not correct. For a set of spectral dithers taken with the S200A1 slit, I see values of [-0.29, -0.42, and -0.54]; the expected values should be roughly [-0.13, 0.0, 0.13], based on the pattern positions [-0.0264, 0, 0.0264] in the ideal frame and a slit width of 0.2 arcsec. SRCYPOS has the correct sequence and values.

It also appears that no wavelength correction is actually being applied regardless of the SRCXPOS value, but I will submit a separate ticket for that.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=544095#comment-544095):_

The dither position is indicated by the _d0* suffix, in this order (horizontal represents the spectral dithers, vertical the primary dithers):

d01 d02 d03

d04 d05 d06

d07 d08 d09

So, d01/d04/d07 have roughly the same SRCXPOS, and likewise for the rest.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=558872#comment-558872):_

[James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) PR #6457 addresses some of the issues we discussed, also mentioned above. There's one remaining issue - the sign of the x position in the slit is opposite of the sign of meta.dither.x_offset. For example,
```java

x_offset = -0.0264 : 
v2, v3 for position #1: 332.8346289845707, -480.061102906561
mx, my in the MSA frame for position #1: 0.0269827684100399, -0.0031764409819799184, 2e-06
sx, sy in the slit frame for position #1: 0.12711058026410074, -0.3616656933616285, 2e-06


xoffset = 0.0264 : 
v2, v3 for position #1: 332.874332560069, -480.02629678507844
mx, my in the MSA frame for position #1: 0.026962084764319105, -0.0031766109926512263, 2e-06
sx, sy in the slit frame for position #1: -0.12714443474453552, -0.3617994368213751, 2e-06

```
When we discussed this I removed "vparity" from the transform from Ideal to V2V3 frame and that "fixed" the sign issue. But I realized that this is incorrect and the parity needs to be taken into account. 

 

So my question is where in the instrument model should the parity be included - is this the OTE reference file (transform)? Since these are two different ways of computing V2, V3 coordinates, one using SIAF transforms, one using the instrument model, it seems to me there's no agreement between them in terms of which way the rotation goes. This is just a guess on my part so it may be completely off.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559192#comment-559192):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) After mulling this over to the point of smoke rising from my head, I think I may have been misinterpreting the definition of x_/y_offset. If those represent the motion of the telescope during the dither, as opposed to the source position, then the sign should be opposite. In that case, vparity should be included in the ideal-v2v3 transform, and we should be ok. Before you proceed, let me verify with the PPS folks, since the values are originally populated in the PPSDB by APT.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559400#comment-559400):_

[Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva)  I confirm that x_/y_offset indicates the telescope motion, so the sign should be opposite that of the source position within an aperture.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559425#comment-559425):_

🎉 🎉 🎉 I hope it's all good this time. Here's the PR

#6457

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559444#comment-559444):_

Fixed by #6457

**Comment by stscijgbot-jp** [comment]
_Comment by [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol) on [JIRA](https://jira.stsci.edu/browse/JP-2103?focusedCommentId=559604#comment-559604):_

Tested on simulated FS data, x and ypos values are now being calculated and stored correctly.
